keras/engine/saving.py

```diff
@@ -333,6 +342,84 @@ def convert_custom_objects(obj):
     return model
 
 
+def _google_storage_transfer(source_filepath, target_filepath, overwrite=True):
+    """Transfers file to/from Google Cloud Storage"""
+    if tf_file_io is None:
+        raise ImportError('Google Storage file transfer requires tensorflow.')
+    if not overwrite:
+        if tf_file_io.file_exists(target_filepath):
+            raise IOError('Object {} already exists.'.format(target_filepath))
+    with tf_file_io.FileIO(source_filepath, mode='r') as source_f:
+        with tf_file_io.FileIO(target_filepath, mode='w') as target_f:
+            target_f.write(source_f.read())
+
+
+def _is_google_storage_location(filepath):
+    """Checks if `filepath` is referencing a google storage bucket"""
+    return isinstance(filepath, string_types) and filepath.startswith('gs://')
+
+
+def parse_save_to_external_resource(save_function):
+    """Function decorator that parses `filepath` argument to the `save_function`
+    and saves the file to the inferred external resource.
+
+    Currently, Google Storage (GS) is supported for filepath:s starting with
+    "gs://".
+    """
+    @wraps(save_function)
+    def save_wrapper(obj, filepath, overwrite=True, *args, **kwargs):
+        if _is_google_storage_location(filepath):
+            tmp_filepath = os.path.join(tempfile.gettempdir(),
+                                        os.path.basename(filepath))
+            save_function(obj, tmp_filepath, True, *args, **kwargs)
+            try:
+                _google_storage_transfer(tmp_filepath, filepath, overwrite)
+            finally:
+                os.remove(tmp_filepath)
+        else:
+            save_function(obj, filepath, *args, **kwargs)
+
+    return save_wrapper
+
+
+def parse_load_from_external_resource(load_function):
+    """Function decorator that parses `filepath` argument to the `load_function`
+    and loads the file from the inferred external resource.
+
+    Currently, Google Storage (GS) is supported for filepath:s starting with
+    "gs://"
+    """
+    def extract_named_arg(f, name, args, kwargs):
+        if name in kwargs:
+            arg = kwargs.pop(name)
+            return arg, args, kwargs
+        argnames = inspect.getargspec(f)[0]
+        for i, (argname, arg) in enumerate(zip(argnames, args)):
+            if argname == name:
+                return arg, args[:i] + args[i + 1:], kwargs
+        else:
+            raise ValueError('function {} has no argument {}'.format(f, name))
+
+    @wraps(load_function)
+    def load_wrapper(*args, **kwargs):
+        filepath, _args, _kwargs = extract_named_arg(
+            load_function, 'filepath', args, kwargs)
+        if _is_google_storage_location(filepath):
+            tmp_filepath = os.path.join(tempfile.gettempdir(),
```

Comment:
In many cases we may not have disk access (e.g. Colab). I think we should attempt to save to a virtual file in memory first, except OOM and then attempt disk. What do you think?

Reply:
Yeah, I looked briefly into that. Seems like this will be made easy when h5py 2.9 is available: http://docs.h5py.org/en/latest/high/file.html#python-file-like-objects Maybe this can be improved in a later PR?

Reply:
@fchollet the following is an option as well:
https://github.com/keras-team/keras/issues/9343#issuecomment-440903847

The upside is that it does not have to wait for h5py 2.9, but the downside is that the code is somewhat hacky and messes with h5py internals.

Reply:
I think we should support it now. Opened this https://github.com/keras-team/keras/pull/11708 - all credits to @obi1kenobi for the internals of the loading part. If there is high-level support in h5py 2.9 we can remove it then.